File: src/main/java/no/nav/syfo/consumer/ws/AktorConsumer.java
package no.nav.syfo.consumer.ws;

import lombok.extern.slf4j.Slf4j;
import no.nav.tjeneste.virksomhet.aktoer.v2.AktoerV2;
import no.nav.tjeneste.virksomhet.aktoer.v2.HentIdentForAktoerIdPersonIkkeFunnet;
import no.nav.tjeneste.virksomhet.aktoer.v2.meldinger.WSHentIdentForAktoerIdRequest;
import org.springframework.stereotype.Component;

import javax.inject.Inject;

import static org.springframework.util.StringUtils.isEmpty;

@Component
@Slf4j
public class AktorConsumer {

    private AktoerV2 aktoerV2;

    @Inject
    public AktorConsumer(AktoerV2 aktoerV2) {
        this.aktoerV2 = aktoerV2;
    }

    public String finnFnr(String aktoerId) {
        if (isEmpty(aktoerId)) {
            throw new RuntimeException("Kan ikke slå opp i AktørId-tjenesten: Har ikke aktør-id.");
        }

        try {
            return aktoerV2.hentIdentForAktoerId(new WSHentIdentForAktoerIdRequest().withAktoerId(aktoerId)).getIdent();
        } catch (HentIdentForAktoerIdPersonIkkeFunnet e) {
            log.error("Fnr ikke funnet", e);
            throw new RuntimeException("Fant ikke fnr for aktørId: " + aktoerId);
        }
    }
}

File: src/main/kotlin/no/nav/syfo/consumer/aktor/AktorConsumer.kt
package no.nav.syfo.consumer.aktor

import no.nav.syfo.config.ApplicationConfig.CALL_ID
import no.nav.syfo.consumer.token.TokenConsumer
import no.nav.syfo.log
import org.slf4j.MDC
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod.GET
import org.springframework.http.HttpStatus.OK
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.RestTemplate
import org.springframework.web.util.UriComponentsBuilder

@Component
class AktorConsumer(private val tokenConsumer: TokenConsumer,
                    @Value("\${srvsyfogsak.username}") private val username: String,
                    @Value("\${aktoerregister.api.v1.url}") private val url: String,
                    private val restTemplate: RestTemplate) {

    val log = log()

    fun getAktorId(fnr: String): String {
        return getIdent(fnr, "AktoerId")
    }

    fun finnFnr(aktorId: String): String {
        return getIdent(aktorId, "NorskIdent")
    }

    private fun getIdent(sokeIdent: String, identgruppe: String): String {
        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_FORM_URLENCODED
        headers.set("Authorization", "Bearer " + tokenConsumer.token.access_token)
        headers.set("Nav-Call-Id", MDC.get(CALL_ID))
        headers.set("Nav-Consumer-Id", username)
        headers.set("Nav-Personidenter", sokeIdent)

        val uriString = UriComponentsBuilder
                .fromHttpUrl("$url/identer")
                .queryParam("gjeldende", "true")
                .queryParam("identgruppe", identgruppe)
                .toUriString()
        try {
            val result = restTemplate
                    .exchange(uriString, GET, HttpEntity<Any>(headers), AktorResponse::class.java)

            if (result.statusCode != OK) {
                val message = "Kall mot aktørregister feiler med HTTP-" + result.statusCode
                log.error(message)
                throw RuntimeException(message)
            }

            return result
                    .body
                    ?.get(sokeIdent)
                    .let { aktor ->
                        aktor?.identer ?: throw RuntimeException("Fant ikke aktøren: " + aktor?.feilmelding)
                    }
                    .filter { ident -> ident.gjeldende }
                    .map { ident -> ident.ident }
                    .first()


        } catch (e: HttpClientErrorException) {
            log.error("Feil ved oppslag i aktørtjenesten", e)
            throw RuntimeException(e)
        }

    }
}

